Cufinufft singles

garrettwrong opened the pull request.

garrettwrong committed:
Bring in cufinufft backend, reproduce basic finufft unit test.

Adds crude adjoint unit test.

garrettwrong committed:
Construct optional 'gpu' feature in setup.py

garrettwrong commented on setup.py:
```diff
@@ -45,6 +45,17 @@ def get_version():
         'scikit-image'
     ],
 
+    # Note, we probably do not want pycuda in the mainline.
```
Note, this actually broke our CI when I left it in earlier. That is perfect example...

junchaoxia replied:
I believe this is required from the original ASPIRE-python package under src/aspire/aspire. It might be better to left it as is before we clean up the code. For the APPLE-Picker module, we use cupy and set a way to switch from numpy to cupy with an option in src/aspire/config.ini. The switch interface is under  src/aspire/utils/numeric/ and the code example is src/aspire/apple/helper.py. It will be great if we can follow the similar way for Cov2D. By the way, the precompiled  pypi package for GPU finufft will be much easy to deal with by using the similar idea as the cpu version of pypi package. Ideally we don't want normal users to get finufft lib by recompiling.

garrettwrong replied:
Thanks for taking a peek.  

Pycuda is clearly not directly installed by the requirements or setup.py, except here where I have just added it.  _We should not assume all users can install or run pycuda._  Yes, Pycuda is used in `src/aspire/aspire/em_classavg/*`, but importing and using it in ASPIRE code without actually having it as a dep is a bug for another day... this partly addressed that (first step). The fact that this isn't caught by the CI systems just means that code is not covered by tests. These are the sort of things I can help with.

I haven't looked at the APPLE-Picker code.  I think if you look at what this code does you might find it is already doing what you describe in the `config.ini`, by adding the `cufinufft` backend :).

I do not think there will ever be precompiled pypi package for CUDA C/C++ code.  It is not a thing that can easily exist because of basic software engineering reasons we talked about yesterday.  Not rehashing here, but we can chat about it again next time if you want.  Even the host C/C++ code is very challenging to python package correctly, and what was done for our py finufft, as I understand it, is not correct or following PyPUG; and this is reflected in that it only works for certain systems (which doesn't include my quite capable Fedora Intel i7 box, or anything in the rack of dell machines in my basement). I don't think it is good to release that to users as it is... I think that was just created for a few users in particular, works for them, and was a big help.  We should continue down that path, but it is not so simple (a different discussion really) to make portable python packages of low level code.  You will find instead a lot of python packaged ""CUDA"" codes are really using pycuda meta programming, and not actually written as CUDA C/C++ programs; this is the case for pynufft.  A different design...

junchaoxia replied:
Yes. I believe the code under  `src/aspire/aspire/em_classavg/` is not merged. I don't think it can run well before we clean up. Somebody just included the whole folder over there. 

For the GPU hackathon, we should build a new branch such as `hackathon2019` and include the code of this PR over there. It will be quite messy after a few people working on it. We don't want the code be merged into the develop branch without cleaning up after the Hackathon.  Namely we should push to hackathon2019 temporarily.

garrettwrong replied:
Not sure what you mean, the `em_classavg` code is definitely merged into the codebase, master and dev.

Yes, we will make such a branch. The hackathon is cute and everything, but it really doesn't have much to do with actual package development or improving the code.  I think of it more as a workshop for people less familiar with GPGPU programming. (Don't get me wrong, that is a good thing).

Most of the prep work for the hackathon _will_ eventually be mergable.  Currently cufinufft should settle just a little but more. This will hang out as a draft.

I am not clear what exactly we plan to do _at_ the hackathon, but that is not related to this PR. As far as this code goes, I am working on the `many` interface right now, and I may have a task for you, similar to your other work, if I don't get it done sooner. we can talk about it soon.

junchaoxia replied:
I did not check that the `em_classavg` code can be run smoothly based on our environment. You can find an environment.yml under the subfolder.  It makes sense to merge this PI before we do GPU Hackathon.

garrettwrong replied:
Okay, I want to update the repo address for cufinufft first. Give me five minutes and I'll push that.  Joakim should also look it over before merge into dev, hackathon or not, he has a good eye for things I miss.

Also if you want to make a adjoint-ness test (I think you were doing this for another code anyway correct?) that would keep the brutish test I made out of the tree. Though this isn't necessary.  One way or another we need a forward and backwards test... I'm not sure I will have time for that before the event myself.

junchaoxia replied:
My implementation of checking adjoint operator has been pushed back to the orientation branch, see the link, https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/orientation/tests/test_PolarBasis2D.py#L211. Let me know if you need me work on it later.

garrettwrong committed:
Add nufft many mode to support hackathon goals.

junchaoxia commented on tests/test_nfft.py:
```diff
@@ -2,53 +2,326 @@
+
+        self.recip_space = np.array([-0.05646675 + 1.503746j, 1.677600 + 0.6610926j, 0.9124417 - 0.7394574j, -0.9136836 - 0.5491410j])
+
+        self.adjoint_vol = np.array([[[9.59495207e-01-1.06291322e+00j,  4.96001394e-01+1.60922425e+00j,
```
This numpy array has more than 250 line. It might be better to use a small size or load it as a data file.

garrettwrong replied:
Yes, I know, its not going to stay.  My opinion is that we shouldn't include bin files if not absolutely necessary.  A better test can be baked without any arrays stored; i just didn't have time to think about 3d.  It wasn't in the way of the hackathon.  In 1d you do something like the following up to scaling:

```
d = vol
dhat = fft(vol)
m = recip
mhat = ifft(recip)
<m,mhat> == <d,dhat>
```
You might figure this out for this case and add commit if you want. That would be helpful :)  @janden might know off his head....

junchaoxia replied:
Previous we have many unit tests including npy data file. I believe we make sure that they are small. Some tests do need data in higher dimensions or many images. Good thing is that we have separated them in a subfolder and it is not necessary for standard delivery. Same things as the tutorials folder.

garrettwrong replied:
That sounds good until you don't know why your repo is 5GB and you can't fix it :D. Whatever you guys like.  I think the mathematical test is more appropriate, but it can be done later.  Text diffs are easier to purge (space), I'd leave it, but Lets see what Joakim wants.  I sort of thought this was all too rough myself....otoh if using dev makes you guys more comfortable it is no difference to me. You can, for example, just branch off this branch... just saying.

junchaoxia replied:
Yes. let's discuss it tomorrow.

janden replied:
> Previous we have many unit tests including npy data file. I believe we make sure that they are small.

Right. This is better than including large binary files, of course, but I think it would be better to avoid hard-coded tests as much as possible. (This is what #156 is about for the block diagonal code.) One reason is size, yes (you're not getting those bytes in the repo back), but another reason is that hard-coded tests are not easy to inspect if something goes wrong. They don't check consistency, but to what extent the code matches with some fixed reference. This may be a more important issue than just size of the tests.

junchaoxia reviewed (commented):
This PR includes a very good draft for the interface to cufinufft.  There are a lots place need to add docstrings in the cufinufft.py file. We should merge it and make it ready for GPU Hackathon in there week.

garrettwrong committed:
update repo addr for cufinufft

cuffinufft was transfered from Melody's repo (and several forks) to permanent home at the Flatiron...

junchaoxia committed:
Add examples for GPU Hackathon 2020

garrettwrong committed:
Some package tertiary deps weren't accurate, this order fixed on traverse

garrettwrong committed:
some tweaks for installing on Ascent

garrettwrong committed:
norm is diff in last digit on power, use allclose

garrettwrong committed:
allow newer scipy so we can use the conda install on ascent

garrettwrong committed:
make finufft optional extra

garrettwrong committed:
typo in setup.py

garrettwrong committed:
add profiler hack

garrettwrong committed:
unsupported character

garrettwrong committed:
Junchao mentioned maybe easier as seperate file

garrettwrong committed:
add example timing loop for use w/ scriptable example

garrettwrong committed:
Merge pull request #159 from ComputationalCryoEM/hackathon_2020_prof

Hackathon 2020 prof

garrettwrong committed:
Disable CI for hack branch

junchaoxia committed:
Merge pull request #160 from ComputationalCryoEM/hackathon_2020_disable_ci

Disable CI for hack branch

garrettwrong committed:
rm vestigial profiling lines

garrettwrong committed:
whitespaces

garrettwrong committed:
Merge branch 'hackathon_2020' into cufinufft_singles

garrettwrong committed:
rm vestigial debug print

garrettwrong committed:
fix merge conflict

junchaoxia committed:
Merge branch 'develop' into hackathon_2020

junchaoxia committed:
Merge branch 'hackathon_2020' into cufinufft_singles and solve conflicts
